mesa/experimental/devs/simulator.py

```diff
@@ -0,0 +1,265 @@
+import numbers
+from typing import Any, Callable
+
+from .eventlist import EventList, Priority, SimulationEvent
+
+
+class Simulator:
+    """The Simulator controls the time advancement of the model.
+
+    The simulator uses next event time progression to advance the simulation time, and execute the next event
+
+    Attributes:
+        event_list (EventList): The list of events to execute
+        time (float | int): The current simulation time
+        time_unit (type) : The unit of the simulation time
+        model (Model): The model to simulate
+
+
+    """
+
+    # TODO: add replication support
+    # TODO: add experimentation support
+
+    def __init__(self, time_unit: type, start_time: int | float):
+        # should model run in a separate thread,
+        # and we can then interact with start, stop, run_until, and step?
+        self.event_list = EventList()
+        self.time = start_time
+        self.time_unit = time_unit
+        self.model = None
+
+    def check_time_unit(self, time: int | float) -> bool:
+        ...
+
+    def setup(self, model: "Model") -> None:
+        """Setup the model to simulate
+
+        Args:
+            model (Model): The model to simulate
+
+        Notes:
+            The basic assumption of the simulator is that a Model has a model.setup method that sets up the
+            model.
+
+        """
+
+        self.event_list.clear()
+        self.model = model
+        model.setup()
+
+    def reset(self):
+        """Reset the simulator by clearing the event list and removing the model to simulate"""
+        self.event_list.clear()
+        self.model = None
+
+    def run(self, time_delta: int | float):
+        """run the simulator for the specified time delta
+
+        Args:
+            time_delta (float| int): The time delta. The simulator is run from the current time to the current time
+                                     plus the time delta
+
+        """
+
+        end_time = self.time + time_delta
+        while self.time < end_time:
+            self.step()
+
+    def step(self):
+        event = self.event_list.pop_event()
+        self.time = event.time
+        event.execute()
+
+    def schedule_event_now(
```

Comment:
Do you think it would improve code quality to make one base function and let the other 2 or 3 depend on it? There is quite some replication between the different event functions.

Schedule now can also just be relative schedule with `0` I think.

Reply:
`schedule_event_now` can indeed be simpler. I don't see any other obvious simplifications